File: docs/palindromes-in-peg/scaffold_circuit_program.py
"""Lower real finite FPP/DP instruction tables to symbolic scaffold equations."""
from fpp_finite import BLANK
from scaffold_circuit import (Value, PREVIOUS, TRUE, FALSE, conjunction,
                              disjunction, neg, choose)
from scaffold_circuit_structs import Tape, StackPool
from read_blocks import analyze


class Program:
  def __init__(self, circuit, kernel, name, quantum=1, extra_symbols=(), coarse=False,
               shared_cells=False, extra_moves=0):
    self.circuit, self.kernel, self.name = circuit, kernel, name
    self.blocks = analyze(kernel) if coarse else None
    alphabets = [{BLANK, *(extra_symbols.get(t, ()) if isinstance(extra_symbols, dict) else extra_symbols)}
                 for t in range(kernel.ntapes)]
    for row in kernel.code:
      if row[0] == "read": alphabets[row[1]].update(row[2])
      if row[0] == "write": alphabets[row[1]].add(row[2])
    if shared_cells and coarse:
      raise ValueError("shared instruction cells require one raw instruction per step")
    self.shared_pool = None
    self.instruction_index, self.quantum = 0, quantum
    if shared_cells:
      self.shared_pool = StackPool(circuit, {"cells": quantum + extra_moves},
                                   tuple(sorted(set().union(*alphabets))))
      # Explicit instruction slots occupy the prefix. Moves requested by the
      # surrounding controller receive distinct slots after that prefix.
      self.shared_pool.used["cells"] = quantum
    self.tapes = [Tape(circuit, f"{name}.t{t}", tuple(sorted(alphabets[t])),
                      slots=quantum if self.blocks is None else
                      tuple(max(1, quantum * n) for n in self.blocks["bounds"][2*t:2*t+2]),
                      pool=self.shared_pool)
                  for t in range(kernel.ntapes)]
    self.pc_key, self.done_key = name + ".pc", name + ".done"
    self.pc = circuit.get(PREVIOUS, self.pc_key, tuple(range(len(kernel.code))), kernel.start)
    self.done = circuit.get(PREVIOUS, self.done_key, (False, True), True).eq(True)

  def start(self, enabled=TRUE, entry=None):
    entry = self.kernel.start if entry is None else entry
    self.pc = Value.select(enabled, Value.constant(entry), self.pc)
    self.done = choose(enabled, FALSE, self.done)

  def reset(self, enabled=TRUE):
    for tape in self.tapes: tape.reset(enabled)
    self.pc = Value.select(enabled, Value.constant(self.kernel.start), self.pc)
    self.done = choose(enabled, TRUE, self.done)

  def step(self, enabled=TRUE):
    if self.blocks is None:
      self.instruction_step(enabled)
    else:
      self.block_step(enabled)

  def instruction_step(self, enabled=TRUE):
    slot = None
    if self.shared_pool is not None:
      if self.instruction_index >= self.quantum:
        raise ValueError("finite instruction cell layout exhausted")
      slot = self.instruction_index
      self.instruction_index += 1
    active = conjunction(enabled, neg(self.done))
    events = [conjunction(active, self.pc.eq(i)) for i in range(len(self.kernel.code))]
    next_pc, halts = {}, []
    writes = [{} for tape in self.tapes]
    moves = [{-1: [], 1: []} for tape in self.tapes]
    for event, row in zip(events, self.kernel.code):
      op = row[0]
      if op == "halt":
        halts.append(event)
      elif op == "read":
        for symbol, target in row[2].items():
          branch = conjunction(event, self.tapes[row[1]].focus.eq(symbol))
          next_pc[target] = disjunction(next_pc.get(target, FALSE), branch)
      elif op in ("write", "move"):
        next_pc[row[3]] = disjunction(next_pc.get(row[3], FALSE), event)
        if op == "write":
          writes[row[1]][row[2]] = disjunction(writes[row[1]].get(row[2], FALSE), event)
        else:
          moves[row[1]][row[2]].append(event)
      else:
        raise ValueError("only finite read/write/move/halt instructions can be lowered")
    for tape, values, directions in zip(self.tapes, writes, moves):
      if values:
        tape.write(Value(values), disjunction(*values.values()))
      tape.move(-1, disjunction(*directions[-1]), slot=slot)
      tape.move(1, disjunction(*directions[1]), slot=slot)
    halt = disjunction(*halts)
    self.pc = Value.select(conjunction(active, neg(halt)), Value(next_pc), self.pc)
    self.done = disjunction(self.done, halt)

  def block_step(self, enabled=TRUE):
    active = conjunction(enabled, neg(self.done))
    events = [conjunction(active, self.pc.eq(i)) for i in range(len(self.kernel.code))]
    next_pc, halts = {}, []
    bases = {(t, side): stack.pool.used.get(stack.name, 0)
             for t, tape in enumerate(self.tapes) for side, stack in enumerate((tape.left, tape.right))}

    def forward(target, event):
      if target in self.blocks["cuts"]:
        next_pc[target] = disjunction(next_pc.get(target, FALSE), event)
      else:
        events[target] = disjunction(events[target], event)

    for state in self.blocks["order"]:
      event, row = events[state], self.kernel.code[state]
      op = row[0]
      if op == "halt":
        halts.append(event)
      elif op == "read":
        for symbol, target in row[2].items():
          forward(target, conjunction(event, self.tapes[row[1]].focus.eq(symbol)))
      else:
        tape = self.tapes[row[1]]
        if op == "write": tape.write(Value.constant(row[2]), event)
        else:
          side = int(row[2] == -1)
          tape.move(row[2], event, slot=bases[row[1], side] + self.blocks["slots"][state])
        forward(row[3], event)
    halt = disjunction(*halts)
    self.circuit.require(disjunction(halt, *next_pc.values()), active)
    self.pc = Value.select(conjunction(active, neg(halt)), Value(next_pc), self.pc)
    self.done = disjunction(self.done, halt)

  def finalize(self):
    for tape in self.tapes: tape.finalize()
    self.circuit.put(self.pc_key, self.pc)
    self.circuit.put(self.done_key, Value.select(self.done, Value.constant(True), Value.constant(False)))

File: scala/pal/src/main/scala/pal/ScaffoldCircuitProgram.scala
package pal

import scala.collection.mutable
import Expr.{TRUE, FALSE}
import Ref.PREVIOUS
import ScaffoldCircuit.*
import FppFinite.{BLANK, Instruction}

/** Lower real finite FPP/DP instruction tables to symbolic scaffold equations.
  * Tape symbols are Strings (as in FppFinite); input heads expose Chars.
  */
object ScaffoldCircuitProgram {
  /** Adapter between the finite kernel and the independent read-block analysis. */
  def readBlockProgram(kernel: FppFinite.Program): ReadBlocks.Program = new ReadBlocks.Program {
    val ntapes: Int = kernel.ntapes
    val code: Vector[ReadBlocks.Instruction] = kernel.instructions.map {
      case Instruction.Halt => ReadBlocks.Instruction.Halt
      case Instruction.Read(tape, choices) => ReadBlocks.Instruction.Read(tape, choices)
      case Instruction.Write(tape, symbol, next) => ReadBlocks.Instruction.Write(tape, symbol, next)
      case Instruction.Move(tape, direction, next) => ReadBlocks.Instruction.Move(tape, direction, next)
      case Instruction.Emit(next) => ReadBlocks.Instruction.Emit(next)
    }
  }

  final class Program(val circuit: Circuit, val kernel: FppFinite.Program, val name: String,
                      val quantum: Int = 1, extraSymbols: Map[Int, Seq[String]] = Map.empty,
                      coarse: Boolean = false, sharedCells: Boolean = false, extraMoves: Int = 0) {
    val blocks: Option[ReadBlocks.Analysis] = if (coarse) { Some(ReadBlocks.analyze(readBlockProgram(kernel))) } else { None }
    private val code = kernel.instructions
    private val alphabets = Array.tabulate(kernel.ntapes)(t => mutable.Set.from(Vector(BLANK) ++ extraSymbols.getOrElse(t, Nil)))
    code.foreach {
      case Instruction.Read(tape, choices) => alphabets(tape) ++= choices.keys
      case Instruction.Write(tape, symbol, _) => alphabets(tape) += symbol
      case _ => ()
    }
    if (sharedCells && coarse) {
      throw new IllegalArgumentException("shared instruction cells require one raw instruction per step")
    }
    val sharedPool: Option[StackPool] = if (sharedCells) {
      val pool = new StackPool(circuit, Vector("cells" -> (quantum + extraMoves)), alphabets.iterator.flatten.toVector.distinct.sorted)
      // Controller moves use distinct slots after the instruction prefix.
      pool.used("cells") = quantum
      Some(pool)
    } else { None }
    var instructionIndex: Int = 0
    val tapes: Vector[Tape] = Vector.tabulate(kernel.ntapes) { t =>
      val alphabet = alphabets(t).toVector.sorted
      blocks match {
        case None => new Tape(circuit, s"$name.t$t", alphabet, quantum, BLANK, sharedPool)
        case Some(b) => Tape.withSides(circuit, s"$name.t$t", alphabet,
          (math.max(1, quantum * b.bounds(2 * t)), math.max(1, quantum * b.bounds(2 * t + 1))), BLANK, sharedPool)
      }
    }
    val pcKey: String = name + ".pc"
    val doneKey: String = name + ".done"
    var pc: Value[Int] = circuit.get(PREVIOUS, pcKey, code.indices.toVector, kernel.start)
    var done: Expr = circuit.get(PREVIOUS, doneKey, Vector(false, true), true).eqTo(true)

    def start(enabled: Expr = TRUE, entry: Option[Int] = None): Unit = {
      pc = Value.select(enabled, Value.constant(entry.getOrElse(kernel.start)), pc)
      done = choose(enabled, FALSE, done)
    }

    def reset(enabled: Expr = TRUE): Unit = {
      tapes.foreach(_.reset(enabled))
      pc = Value.select(enabled, Value.constant(kernel.start), pc)
      done = choose(enabled, TRUE, done)
    }

    def step(enabled: Expr = TRUE): Unit = {
      if (blocks.isEmpty) { instructionStep(enabled) } else { blockStep(enabled) }
    }

    def instructionStep(enabled: Expr = TRUE): Unit = {
      val slot = sharedPool.map { _ =>
        if (instructionIndex >= quantum) { throw new IllegalArgumentException("finite instruction cell layout exhausted") }
        val index = instructionIndex
        instructionIndex += 1
        index
      }
      val active = conjunction(enabled, neg(done))
      val events = code.indices.map(i => conjunction(active, pc.eqTo(i)))
      val nextPc = mutable.LinkedHashMap.empty[Int, Expr]
      val halts = mutable.ArrayBuffer.empty[Expr]
      val writes = Array.fill(tapes.size)(mutable.LinkedHashMap.empty[String, Expr])
      val moves = Array.fill(tapes.size)(Map(-1 -> mutable.ArrayBuffer.empty[Expr], 1 -> mutable.ArrayBuffer.empty[Expr]))
      def forward(target: Int, event: Expr): Unit = { nextPc(target) = disjunction(nextPc.getOrElse(target, FALSE), event) }
      for ((event, row) <- events.zip(code)) {
        row match {
          case Instruction.Halt => halts += event
          case Instruction.Read(tape, choices) =>
            choices.foreach { case (symbol, target) => forward(target, conjunction(event, tapes(tape).focus.eqTo(symbol))) }
          case Instruction.Write(tape, symbol, target) =>
            forward(target, event)
            writes(tape)(symbol) = disjunction(writes(tape).getOrElse(symbol, FALSE), event)
          case Instruction.Move(tape, direction, target) =>
            forward(target, event)
            moves(tape)(direction) += event
          case Instruction.Emit(_) =>
            throw new IllegalArgumentException("only finite read/write/move/halt instructions can be lowered")
        }
      }
      for (t <- tapes.indices) {
        if (writes(t).nonEmpty) { tapes(t).write(Value(writes(t)), disjunction(writes(t).values.toSeq*)) }
        tapes(t).move(-1, disjunction(moves(t)(-1).toSeq*), slot)
        tapes(t).move(1, disjunction(moves(t)(1).toSeq*), slot)
      }
      val halt = disjunction(halts.toSeq*)
      pc = Value.select(conjunction(active, neg(halt)), Value(nextPc), pc)
      done = disjunction(done, halt)
    }

    def blockStep(enabled: Expr = TRUE): Unit = {
      val analysis = blocks.getOrElse(throw new IllegalStateException("read-block analysis required"))
      val active = conjunction(enabled, neg(done))
      val events = code.indices.map(i => conjunction(active, pc.eqTo(i))).toArray
      val nextPc = mutable.LinkedHashMap.empty[Int, Expr]
      val halts = mutable.ArrayBuffer.empty[Expr]
      val bases = tapes.map(tape => Vector(tape.left, tape.right).map(stack => stack.pool.used.getOrElse(stack.name, 0)))
      def forward(target: Int, event: Expr): Unit = {
        if (analysis.cuts(target)) { nextPc(target) = disjunction(nextPc.getOrElse(target, FALSE), event) }
        else { events(target) = disjunction(events(target), event) }
      }
      for (state <- analysis.order) {
        val event = events(state)
        code(state) match {
          case Instruction.Halt => halts += event
          case Instruction.Read(tape, choices) =>
            choices.foreach { case (symbol, target) => forward(target, conjunction(event, tapes(tape).focus.eqTo(symbol))) }
          case Instruction.Write(tape, symbol, target) =>
            tapes(tape).write(Value.constant(symbol), event)
            forward(target, event)
          case Instruction.Move(tape, direction, target) =>
            val side = if (direction == -1) { 1 } else { 0 }
            tapes(tape).move(direction, event, Some(bases(tape)(side) + analysis.slots(state)))
            forward(target, event)
          case Instruction.Emit(_) => throw new IllegalArgumentException("read blocks require read/write/move/halt instructions")
        }
      }
      val halt = disjunction(halts.toSeq*)
      circuit.require(disjunction((halt +: nextPc.values.toVector)*), active)
      pc = Value.select(conjunction(active, neg(halt)), Value(nextPc), pc)
      done = disjunction(done, halt)
    }

    def commit(): Unit = {
      tapes.foreach(_.commit())
      circuit.put(pcKey, pc)
      circuit.put(doneKey, Value.select(done, Value.constant(true), Value.constant(false)))
    }
  }
}
